### PR TITLE
docs: how to control GPIO pins within rootless Docker container on Ra…

### PR DESCRIPTION
**Pull request by mu88**
Relates to #2169 

Dear @Ellerbach,

I kindly ask you for your feedback regarding the new page "Control GPIO pins within rootless Docker container on Raspberry Pi". 

Since I'm not a native English speaker, I also appreciate feedback regarding certain sentences/words that might sound "bumpy" 🙂

Last but not least, maybe we can consider whether it's worth asking somebody from the "rootless Docker .NET" world like Rich Landers for his feedback 🤔 just to be on the safe side that I'm not writing anything wrong.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/iot/pull/2276)

### CODE REVIEW

**Comment by mu88** [comment]
@dotnet-policy-service agree

**Comment by Ellerbach** [comment]
azp /run

File: Documentation/raspi-Docker-GPIO.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+## Running the app in Docker
+
+Assuming we're on a Raspberry Pi, we could now create a container from the previously created image like this:
+```shell
+docker run --rm --name blinkled blinkledsample:latest
+```
+
+However, this will fail as the app has no access to the GPIO pins inside the container. As with every other resource necessary for running a container, it must be mounted from the host into the container. In case of GPIO, the device `/dev/gpiomem` is needed.  
+The complete command to run the container therefore is:
+```shell
```
Comment (Ellerbach) [inline Documentation/raspi-Docker-GPIO.md:93]:
```suggestion

```shell
```

File: Documentation/raspi-Docker-GPIO.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+docker run --rm --device=/dev/gpiomem --name blinkled blinkledsample:latest
+```
+
+But executing the command leads to the following error:
+```
```
Comment (Ellerbach) [inline Documentation/raspi-Docker-GPIO.md:98]:
```suggestion

```text
```

File: Documentation/raspi-Docker-GPIO.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+   at System.Device.Gpio.Drivers.RaspberryPi3LinuxDriver.Initialize()
+   at System.Device.Gpio.Drivers.RaspberryPi3LinuxDriver.OpenPin(Int32 pinNumber)
+   at System.Device.Gpio.Drivers.RaspberryPi3Driver.OpenPin(Int32 pinNumber)
+   at System.Device.Gpio.GpioController.OpenPinCore(Int32 pinNumber)
+   at System.Device.Gpio.GpioController.OpenPin(Int32 pinNumber)
+   at System.Device.Gpio.GpioController.OpenPin(Int32 pinNumber, PinMode mode)
+```
+
+This piece is a bit more tricky: first of all, let's check the permissions inside the container for the device:
+```shell
```
Comment (Ellerbach) [inline Documentation/raspi-Docker-GPIO.md:109]:
```suggestion

```shell
```

File: Documentation/raspi-Docker-GPIO.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+myUser@myRaspi:~ $ docker run -it --rm --entrypoint /bin/bash blinkledsample:latest
+app@18c678710307:/app$ ls -lh /dev/gpiomem
+crw-rw---- 1 root 997 245, 0 Jan 28 18:45 /dev/gpiomem
+```
+Now we know the following about `/dev/gpiomem`:
```
Comment (Ellerbach) [inline Documentation/raspi-Docker-GPIO.md:114]:
```suggestion

Now we know the following about `/dev/gpiomem`:

```

File: Documentation/raspi-Docker-GPIO.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+- It is user-owned by `root` with read and write permissions.
+- It is group-owned by the group with ID `997` with read and write permissions.
+- All others have no access.
+
+So let's test if our container can access the device:
+```shell
```
Comment (Ellerbach) [inline Documentation/raspi-Docker-GPIO.md:120]:
```suggestion

```shell
```

File: Documentation/raspi-Docker-GPIO.md [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+app@18c678710307:/app$ test -r /dev/gpiomem; echo "$?"
+1
+app@18c678710307:/app$ test -w /dev/gpiomem; echo "$?"
+1
+app@18c678710307:/app$ id
+uid=1654(app) gid=1654(app) groups=1654(app)
+```
+We see that the current user `app` can neither read nor write `/dev/gpiomem`, and that the current user  has the group ID `gid=1654`.
```
Comment (Ellerbach) [inline Documentation/raspi-Docker-GPIO.md:128]:
```suggestion

We see that the current user `app` can neither read nor write `/dev/gpiomem`, and that the current user  has the group ID `gid=1654`.  
```

File: Documentation/raspi-Docker-GPIO.md [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+app@18c678710307:/app$ test -w /dev/gpiomem; echo "$?"
+1
+app@18c678710307:/app$ id
+uid=1654(app) gid=1654(app) groups=1654(app)
+```
+We see that the current user `app` can neither read nor write `/dev/gpiomem`, and that the current user  has the group ID `gid=1654`.  
+That perfectly explains the error we're seeing in our app: our app runs as user `app` which belongs to the group `1654`, but only members of the group `997` can access the GPIO device.
+
+Luckily, the desired group ID can be set when running a Docker container:
+```shell
```
Comment (Ellerbach) [inline Documentation/raspi-Docker-GPIO.md:132]:
```suggestion

```shell
```

File: Documentation/raspi-Docker-GPIO.md [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+docker run --rm -u "1654:997" --device=/dev/gpiomem --name blinkled blinkledsample:latest
+```
+
+Now you should see the console output `Blinking LED. Press Ctrl+C to end.` and a blinking LED - congratulations! 🎉
+
+## `root` vs. `app`
```
Comment (Ellerbach) [inline Documentation/raspi-Docker-GPIO.md:138]:
```suggestion
## `root` vs. `app`

```

File: Documentation/raspi-Docker-GPIO.md [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+Starting with .NET 8, the Docker images for Linux come with a non-root user. There is [this excellent blog article][1] about the topic, but it's worth spending a word on the rootless nature of the described containerization.
+
+Without configuring anything manually, the created Docker image uses the new rootless `app` user. That happened automatically since we used the .NET SDK Container Building Tools.  
+
+In case of using a dedicated `Dockerfile` and building the image via `docker build`, the following line would have to be added:
+```dockerfile
```
Comment (Ellerbach) [inline Documentation/raspi-Docker-GPIO.md:144]:
```suggestion

```dockerfile
```

File: Documentation/raspi-Docker-GPIO.md [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+USER $APP_UID
+```
+This way, the user ID can be controlled via the environment variable `APP_UID` and uses `1654` by default.
```
Comment (Ellerbach) [inline Documentation/raspi-Docker-GPIO.md:147]:
```suggestion

This way, the user ID can be controlled via the environment variable `APP_UID` and uses `1654` by default.  
```

File: Documentation/raspi-Docker-GPIO.md [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+
+Without configuring anything manually, the created Docker image uses the new rootless `app` user. That happened automatically since we used the .NET SDK Container Building Tools.  
+
+In case of using a dedicated `Dockerfile` and building the image via `docker build`, the following line would have to be added:
+```dockerfile
+USER $APP_UID
+```
+This way, the user ID can be controlled via the environment variable `APP_UID` and uses `1654` by default.  
+Not adding the line and omit using `-u "1654:997"` when starting the Docker container will result in using the `root` user, i. e. there will still be a blinking LED, but the app will have more privileges due to the root permissions.
+
```
Comment (Ellerbach) [inline Documentation/raspi-Docker-GPIO.md:149]:
```suggestion
```

File: Documentation/raspi-Docker-GPIO.md [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+
+## Additional information
+
+The following issues and articles provide more in-depth details:
```
Comment (Ellerbach) [inline Documentation/raspi-Docker-GPIO.md:153]:
```suggestion
The following issues and articles provide more in-depth details:

```

**Review by Ellerbach — APPROVED** [review]
Thanks, this is really useful for people in the same situation as you. I made couple of suggestions while reading to have you gaining time with the linter ;-)
Otherwise, as already exchange on the topic in the issue, things look clear to me.

File: Documentation/raspi-Docker-GPIO.md [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="System.Device.Gpio" Version="3.1.0" />
+  </ItemGroup>
+</Project>
+```
+
+As you can see, it is a .NET 8 console app which references the NuGet package `System.Device.Gpio`.
+
+First of all, let's make sure that this compiles nicely by running `dotnet build`:
```
Comment (Ellerbach) [inline Documentation/raspi-Docker-GPIO.md:49]:
```suggestion
First of all, let's make sure that this compiles nicely by running `dotnet build`:

```

**Review by Ellerbach — COMMENTED** [review]
Please check the linter result and check if anything is left!

File: Documentation/raspi-Docker-GPIO.md [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+When it comes to controlling the GPIO pins of a Raspberry Pi, the existing tutorials will get you started in a couple of minutes. But most of the code assumes that it runs on bare metal and has sufficient privileges, e. g. to access the GPIO pins.  
+But what if you want to run this code in a Docker container, or even rootless? This page will explain how to set up all the necessary pieces.
+
+## Prerequisites
+
+In order to follow the instructions below, Docker must be installed on the Raspberry Pi. The official Docker documentation contains an excellent starting point for installing Docker either for the 32-bit (`armhf`, [see here][5]) or 64-bit (`arm64`, [see here][6]) OS.
+
+## Containerizing the app
+
+Let's use the following sample code to blink an LED ([taken from the official documentation][2]) and save it under `BlinkLedSample\Program.cs`:
```
Comment (Ellerbach) [inline Documentation/raspi-Docker-GPIO.md:12]:
```suggestion
Let's use the following sample code to blink an LED ([taken from the official documentation][2]) and save it under `BlinkLedSample\Program.cs`:

```

File: Documentation/raspi-Docker-GPIO.md [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+- [Blink an LED][2]
+- [Containerize a .NET app with dotnet publish][3]
+- [RaspiFanController][4]
+
+[1]: https://devblogs.microsoft.com/dotnet/securing-containers-with-rootless
+[2]: https://learn.microsoft.com/en-us/dotnet/iot/tutorials/blink-led
+[3]: https://learn.microsoft.com/en-us/dotnet/core/docker/publish-as-container
+[4]: https://github.com/mu88/RaspiFanController
+[5]: https://docs.docker.com/engine/install/raspberry-pi-os/
+[6]: https://docs.docker.com/engine/install/debian/
```
Comment (Ellerbach) [inline Documentation/raspi-Docker-GPIO.md:184]:
```suggestion
[6]: https://docs.docker.com/engine/install/debian/

```

**Comment by mu88** [comment]
Is there anything that I can do regarding the failed builds?

File: Documentation/raspi-Docker-GPIO.md (hunk 13 above)
Comment (mu88) [reply Documentation/raspi-Docker-GPIO.md:50]: Thx for solving the various linter issues!

**Comment by Ellerbach** [comment]
> Is there anything that I can do regarding the failed builds?

From the build, I can see 1 broken link in one of the file. Can you just remove it as it doesn't really seems to exist anymore? Bonus point if you find a replacement link. Then good to go!

**Comment by mu88** [comment]
Is there anything I can help you with regarding the failing builds, @Ellerbach ?

**Pull request merged**